### PR TITLE
[scroll-animations] implement parsing support for scroll-timeline and its longhand properties

### PR DESCRIPTION
**Pull request by graouts**
#### 78224352916760e6e360c3e1ad5e981f77aaec7d
<pre>
[scroll-animations] implement parsing support for scroll-timeline and its longhand properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=265302">https://bugs.webkit.org/show_bug.cgi?id=265302</a>

Reviewed by Tim Nguyen.

Add parsing support for the `scroll-timeline` shorthand and its two coordinating list base
property longhands `scroll-timeline-name` and `scroll-timeline-axis` as specified in
<a href="https://drafts.csswg.org/scroll-animations-1/#scroll-timelines-named.">https://drafts.csswg.org/scroll-animations-1/#scroll-timelines-named.</a>

Two WPT failures remain for `scroll-timeline` parsing and a WPT issue has been filed to determine
whether the test is correct: <a href="https://github.com/web-platform-tests/wpt/issues/43336.">https://github.com/web-platform-tests/wpt/issues/43336.</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-axis-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-axis-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-name-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-name-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-shorthand-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::create):
(WebCore::ScrollTimeline::ScrollTimeline):
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForScrollTimelineAxis):
(WebCore::valueForScrollTimelineName):
(WebCore::scrollTimelineShorthandValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
* Source/WebCore/css/ShorthandSerializer.cpp:
(WebCore::ShorthandSerializer::serialize):
(WebCore::ShorthandSerializer::serializeLayered const):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
(WebCore::CSSPropertyParser::consumeScrollTimelineShorthand):
(WebCore::CSSPropertyParser::parseShorthand):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::adjustScrollTimelines):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::initialScrollTimelineAxes):
(WebCore::RenderStyle::initialScrollTimelineNames):
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::scrollTimelines const):
(WebCore::RenderStyle::scrollTimelineAxes const):
(WebCore::RenderStyle::scrollTimelineNames const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setScrollTimelineAxes):
(WebCore::RenderStyle::setScrollTimelineNames):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertScrollTimelineName):
(WebCore::Style::BuilderConverter::convertScrollTimelineAxis):

Canonical link: <a href="https://commits.webkit.org/271110@main">https://commits.webkit.org/271110@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/02bfddb3e06c8ccd5d65aa2d911e312503e1608c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27351 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5990 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28599 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29577 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25032 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27819 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7899 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3389 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24840 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27613 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4788 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23477 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4166 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4308 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24473 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30213 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24987 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24893 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30455 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4336 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2478 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28414 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5800 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4793 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3534 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4717 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/WebCore.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -2665,7 +2665,7 @@
 		7132445120109DA500AE7FB2 /* WebAnimationUtilities.h in Headers */ = {isa = PBXBuildFile; fileRef = 7132444F20109D9B00AE7FB2 /* WebAnimationUtilities.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		7134496E146941B300720312 /* SVGLengthContext.h in Headers */ = {isa = PBXBuildFile; fileRef = 7134496C146941B300720312 /* SVGLengthContext.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		713516932AE6D01F00718EBE /* ScrollTimeline.h in Headers */ = {isa = PBXBuildFile; fileRef = 713516912AE6D00A00718EBE /* ScrollTimeline.h */; settings = {ATTRIBUTES = (Private, ); }; };
-		713516982AE6D2AE00718EBE /* ScrollAxis.h in Headers */ = {isa = PBXBuildFile; fileRef = 713516962AE6D29600718EBE /* ScrollAxis.h */; };
+		713516982AE6D2AE00718EBE /* ScrollAxis.h in Headers */ = {isa = PBXBuildFile; fileRef = 713516962AE6D29600718EBE /* ScrollAxis.h */; settings = {ATTRIBUTES = (Private, ); }; };
```
Comment (mdubet) [inline Source/WebCore/WebCore.xcodeproj/project.pbxproj]: Is this change on purpose?

File: Source/WebCore/style/StyleAdjuster.cpp [hunk 2]
```diff
@@ -610,6 +610,9 @@ void Adjuster::adjust(RenderStyle& style, const RenderStyle* userAgentAppearance
     style.adjustAnimations();
     style.adjustTransitions();
 
+    // Do the same for scroll-timeline longhands.
```
Comment (mdubet) [inline Source/WebCore/style/StyleAdjuster.cpp:613]: // FIXME ?

File: Source/WebCore/WebCore.xcodeproj/project.pbxproj (hunk 1 above)
Comment (graouts) [reply Source/WebCore/WebCore.xcodeproj/project.pbxproj]: Yes, this addresses a compilation error (which seems to happen on non-Apple ports, so something I’ll need to fix).

File: Source/WebCore/style/StyleAdjuster.cpp (hunk 2 above)
Comment (graouts) [reply Source/WebCore/style/StyleAdjuster.cpp:613]: Not a fixme, this comment makes sense in the context of code further up.

File: LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt [hunk 3]
```diff
@@ -288,6 +288,8 @@ PASS scroll-padding-top
 PASS scroll-snap-align
 PASS scroll-snap-stop
 PASS scroll-snap-type
+FAIL scroll-timeline-axis assert_equals: expected "none" but got "block"
```
Comment (nt1m) [inline LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt]: This suggests a problem with `scroll-timeline-axis`, `none` is not a valid value.

Other test regressions are similar.

File: LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-shorthand-expected.txt [hunk 4]
```diff
@@ -1,52 +1,52 @@
...
+PASS e.style['scroll-timeline'] = "--abc y, --def" should set scroll-timeline-name
+PASS e.style['scroll-timeline'] = "--abc y, --def" should not set unrelated longhands
+PASS e.style['scroll-timeline'] = "--abc, --def" should set scroll-timeline-axis
+PASS e.style['scroll-timeline'] = "--abc, --def" should set scroll-timeline-name
+PASS e.style['scroll-timeline'] = "--abc, --def" should not set unrelated longhands
+PASS Shorthand contraction of scroll-timeline-name:--abc:undefined;scroll-timeline-axis:inline:undefined
+PASS Shorthand contraction of scroll-timeline-name:--a, --b:undefined;scroll-timeline-axis:inline, block:undefined
+PASS Shorthand contraction of scroll-timeline-name:none, none:undefined;scroll-timeline-axis:block, block:undefined
+FAIL Shorthand contraction of scroll-timeline-name:--a, --b, --c:undefined;scroll-timeline-axis:inline, inline:undefined assert_equals: Declared value expected "" but got "--a inline, --b inline, --c"
+FAIL Shorthand contraction of scroll-timeline-name:--a, --b:undefined;scroll-timeline-axis:inline, inline, inline:undefined assert_equals: Declared value expected "" but got "--a inline, --b inline, none inline"
```
Comment (nt1m) [inline LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-shorthand-expected.txt:51]: The WPT suggests that when the name & axis have different lengths, the shorthand should be an empty string. Seems reasonable behavior if the shorthand can't express that by itself.

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 5]
```diff
@@ -2820,6 +2821,53 @@ static inline bool isFlexOrGridItem(RenderObject* renderer)
     return box.isFlexItem() || box.isGridItem();
 }
 
+static Ref<CSSValue> valueForScrollTimelineAxis(const Vector<ScrollAxis>& axes)
+{
+    if (axes.isEmpty())
+        return CSSPrimitiveValue::create(CSSValueNone);
```
Comment (nt1m) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]:
```suggestion
    if (axes.isEmpty())
        return CSSPrimitiveValue::create(CSSValueBlock);
```

`none` isn't valid for `scroll-timeline-axis`

File: Source/WebCore/css/parser/CSSPropertyParser.cpp [hunk 6]
```diff
@@ -2626,6 +2630,36 @@ bool CSSPropertyParser::consumeWhiteSpaceShorthand(bool important)
...
+
+    do {
+        // A valid scroll-timeline-name is required.
+        if (auto name = CSSPropertyParsing::consumeSingleScrollTimelineName(m_range))
+            namesList.append(name.releaseNonNull());
+        else
+            return false;
+
+        // A scroll-timeline-axis is optional.
+        if (m_range.peek().type() != IdentToken)
```
Comment (nt1m) [inline Source/WebCore/css/parser/CSSPropertyParser.cpp]:
```suggestion
        if (m_range.peek().type() == CommaToken)
```
I'd be more specific here.

File: Source/WebCore/rendering/style/RenderStyle.cpp [hunk 7]
```diff
@@ -3880,6 +3881,47 @@ UserSelect RenderStyle::effectiveUserSelect() const
...
+
+void RenderStyle::setScrollTimelineAxis(const Vector<ScrollAxis>& axes)
+{
+    m_nonInheritedData.access().rareData.access().scrollTimelineAxes = axes;
+}
+
+void RenderStyle::setScrollTimelineName(const Vector<AtomString>& names)
+{
+    m_nonInheritedData.access().rareData.access().scrollTimelineNames = names;
+}
```
Comment (nt1m) [inline Source/WebCore/rendering/style/RenderStyle.cpp]: Can we use the plural names on RenderStyle? You can use `"name-for-methods"` in `CSSProperties.json` if you want to link them to each other.

Comment (nt1m) [reply Source/WebCore/rendering/style/RenderStyle.cpp]: Also should these methods be in `RenderStyleInlines.h` and `RenderStyleSetters.h` respectively? (No strong opinion here)

File: Source/WebCore/rendering/style/RenderStyle.h [hunk 8]
```diff
@@ -1976,6 +1987,9 @@ class RenderStyle {
     static ScrollSnapAlign initialScrollSnapAlign();
     static ScrollSnapStop initialScrollSnapStop();
 
+    static Vector<ScrollAxis> initialScrollTimelineAxis() { return { }; }
```
Comment (nt1m) [inline Source/WebCore/rendering/style/RenderStyle.h]:
```suggestion
    static Vector<ScrollAxis> initialScrollTimelineAxis() { return { ScrollAxis::Block }; }
```

I think this was correct fwiw.

Comment (nt1m) [reply Source/WebCore/rendering/style/RenderStyle.h]: `{ }` is also correct fwiw, just depends on how you handle serialization of `{ }`. As long as it serializes to `CSSValueBlock` that's fine.

File: Source/WebCore/css/parser/CSSPropertyParser.cpp (hunk 6 above)
Comment (nt1m) [reply Source/WebCore/css/parser/CSSPropertyParser.cpp]: `if (m_range.peek().type() == CommaToken || m_range.atEnd())` is probably the condition you want.

File: Source/WebCore/css/parser/CSSPropertyParser.cpp [hunk 9]
```diff
@@ -2626,6 +2630,36 @@ bool CSSPropertyParser::consumeWhiteSpaceShorthand(bool important)
...
+            axesList.append(CSSPrimitiveValue::create(CSSValueID::CSSValueBlock));
+        else if (auto axis = CSSPropertyParsing::consumeAxis(m_range))
+            axesList.append(axis.releaseNonNull());
+        else
+            return false;
+    } while (consumeCommaIncludingWhitespace(m_range));
+
+    if (namesList.isEmpty())
+        return false;
```
Comment (nt1m) [inline Source/WebCore/css/parser/CSSPropertyParser.cpp:2655]: Do we also need to filter out any junk? e.g. is `scroll-timeline: none junk` rejected? (we should have `test_invalid_value` tests for that)

If we do need to filter out junk, then `if (namesList.isEmpty() || !m_range.atEnd())` is probably the condition you want.

File: LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-shorthand-expected.txt (hunk 4 above)
Comment (graouts) [reply LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-shorthand-expected.txt:51]: I think this is wrong, see https://github.com/web-platform-tests/wpt/issues/43336. This will be addressed by a separate patch which will update the WPT and contain any required source change to match the new expectations.

File: Source/WebCore/css/parser/CSSPropertyParser.cpp (hunk 9 above)
Comment (graouts) [reply Source/WebCore/css/parser/CSSPropertyParser.cpp:2655]: I believe this is tested, there's a test for `--abc --abc` which is correctly rejected.

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 5 above)
Comment (graouts) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: Thanks for catching this!

File: Source/WebCore/css/parser/CSSPropertyParser.cpp (hunk 6 above)
Comment (graouts) [reply Source/WebCore/css/parser/CSSPropertyParser.cpp]: That does seem more comprehensive, I'll adopt it in the next iteration of this patch.

File: Source/WebCore/rendering/style/RenderStyle.h (hunk 8 above)
Comment (graouts) [reply Source/WebCore/rendering/style/RenderStyle.h]: Yeah, this is handled correctly as it stands.

File: Source/WebCore/rendering/style/RenderStyle.cpp (hunk 7 above)
Comment (graouts) [reply Source/WebCore/rendering/style/RenderStyle.cpp]: Yeah, I think the plural name would be nicer, I'll set this up. I'll also look at `RenderStyleSetters.h`. Not sure what qualifies to be in `RenderStyleInlines.h`.

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 10]
```diff
@@ -2820,6 +2821,53 @@ static inline bool isFlexOrGridItem(RenderObject* renderer)
...
+        auto& name = timeline->name();
+        auto axis = timeline->axis();
+
+        ASSERT(!name.isNull());
+        auto nameCSSValue = CSSPrimitiveValue::createCustomIdent(name);
+
+        if (axis == ScrollAxis::Block)
+            list.append(WTFMove(nameCSSValue));
+        else
+            list.append(CSSValueList::createSpaceSeparated(nameCSSValue, createConvertingToCSSValueID(axis)));
```
Comment (nt1m) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]:
```suggestion
            list.append(CSSValuePair::createNonCoalescing(nameCSSValue, createConvertingToCSSValueID(axis)));
```
You can use CSSValuePair here.

File: LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt
Comment (nt1m) [inline LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:1]: You probably want to adjust the baselines for ipad / ios-wk2 / gtk / wpe too.

**Review by nt1m — APPROVED** [review]
